apps/passport-server/src/database/queries/telegram/insertTelegramConversation.ts

```diff
@@ -71,3 +71,55 @@ set topic_name = $3, is_anon_topic = $4;`,
   );
   return result.rowCount;
 }
+
+export async function insertTelegramNullifier(
+  client: Pool,
+  nullifierHash: string,
+  telegramChatId: number,
+  topicId: number,
+  maxMessagesPerDay: number
+): Promise<boolean> {
+  // Get the current usage count for the nullifier.
+  const nullifierTimesUsed = await sqlQuery(
+    client,
+    `\
+    select nullifier_times_used
+    from telegram_chat_anon_nullifiers
+    where nullifier = $1;`,
+    [nullifierHash]
+  );
+
+  const timesUsed = nullifierTimesUsed.rows[0]?.nullifier_times_used;
+
+  const canInsertOrUpdate =
+    timesUsed == undefined || timesUsed < maxMessagesPerDay;
+
+  if (canInsertOrUpdate) {
+    await doUpdateNullifier(client, nullifierHash, telegramChatId, topicId);
+  }
+
+  return canInsertOrUpdate;
+}
+
+async function doUpdateNullifier(
+  client: Pool,
+  nullifierHash: string,
+  telegramChatId: number,
+  topicId: number
+): Promise<void> {
+  const query = `
+    insert into telegram_chat_anon_nullifiers (nullifier, telegram_chat_id, topic_id, nullifier_times_used, last_used_timestamp)
+    values ($1, $2, $3, 1, now())
+    on conflict (nullifier) 
+    do update set
+        nullifier_times_used = 
+          case 
+              when excluded.last_used_timestamp <= now() - interval '24 hours' then 0 
+              else telegram_chat_anon_nullifiers.nullifier_times_used + 1
+          end,
+        last_used_timestamp = now()
+    returning nullifier_times_used;
```

Comment:
I don't think this rate limiting works as intended. For example, I should be allowed to send a message at 8am, 12pm, 4pm, and then 8:01am the following day. With this scheme, I would be blocked when trying to send a message at 8:01am the following day because my last_used_timestamp is 4pm (within 24 hours of 8:01am) and nullifier_times_used is 3.

We shouldn't just be storing the last_used_timestamp, we need *all* timestamps for the last three usages. I would look at `account_reset_timestamps` as an example. Another resource would be looking at [sliding window rate limiters](https://arpitbhayani.me/blogs/sliding-window-ratelimiter/).